pkg/registry/resource/resourceclaim/strategy.go

```diff
@@ -220,3 +224,53 @@ func draAdminAccessFeatureInUse(claim *resource.ResourceClaim) bool {
 
 	return false
 }
+
+func dropDisabledDRAResourceClaimDeviceStatusFields(newClaim, oldClaim *resource.ResourceClaim) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DRAResourceClaimDeviceStatus) {
+		newClaim.Status.Devices = nil
+	}
+}
+
+// dropDeallocatedStatusDevices removes the status.devices that were allocated
+// in the oldClaim and that have been removed in the newClaim.
+func dropDeallocatedStatusDevices(newClaim, oldClaim *resource.ResourceClaim) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DRAResourceClaimDeviceStatus) {
+		return
+	}
+
+	deallocatedDevices := sets.New[structured.DeviceID]()
+
+	if oldClaim.Status.Allocation != nil {
+		// Get all devices in the oldClaim.
+		for _, result := range oldClaim.Status.Allocation.Devices.Results {
+			deviceID := structured.DeviceID{Driver: result.Driver, Pool: result.Pool, Device: result.Device}
+			deallocatedDevices.Insert(deviceID)
+		}
+	}
+
+	// Remove devices from deallocatedDevices that are still in newClaim.
+	if newClaim.Status.Allocation != nil {
+		for _, result := range newClaim.Status.Allocation.Devices.Results {
+			deviceID := structured.DeviceID{Driver: result.Driver, Pool: result.Pool, Device: result.Device}
+			deallocatedDevices.Delete(deviceID)
+		}
+	}
+
+	// Remove from newClaim.Status.Devices.
+	for i := len(newClaim.Status.Devices) - 1; i >= 0; i-- {
+		deviceID := structured.DeviceID{
+			Driver: newClaim.Status.Devices[i].Driver,
+			Pool:   newClaim.Status.Devices[i].Pool,
+			Device: newClaim.Status.Devices[i].Device,
+		}
+		// Device was in the oldClaim.Status.Allocation.Devices but is no longer in the
+		// newClaim.Status.Allocation.Devices so it must be removed from the newClaim.Status.Devices.
+		if deallocatedDevices.Has(deviceID) {
+			newClaim.Status.Devices = append(newClaim.Status.Devices[:i], newClaim.Status.Devices[i+1:]...)
+		}
+	}
```

Comment:
I think you can filter in place https://go.dev/wiki/SliceTricks
```
n := 0
for _, device := range newClaim.Status.Devices {
   		deviceID := structured.DeviceID{
			Driver: newClaim.Status.Devices[i].Driver,
			Pool:   newClaim.Status.Devices[i].Pool,
			Device: newClaim.Status.Devices[i].Device,
		}
                if !deallocatedDevices.Has(deviceID) {
                           newClaim.Status.Devices[n] = device
                           n++
                 }
}
newClaim.Status.Devices  = newClaim.Status.Devices[:n]
```

Reply:
I don't have preference here, I took @aojea solution, I think it's quite clean and easy to understand code.